api/data_explorer/__main__.py

```diff
@@ -199,26 +207,43 @@ def _process_facets(es):
     app.app.config['NESTED_PATHS'] = elasticsearch_util.get_nested_paths(es)
 
     for facet_config in facets_config:
-        es_field_name = facet_config['elasticsearch_field_name']
-        if es_field_name in facets:
-            raise EnvironmentError('%s appears more than once in ui.json' %
-                                   es_field_name)
-        field_type = elasticsearch_util.get_field_type(es, es_field_name)
-        ui_facet_name = facet_config['ui_facet_name']
-        if es_field_name.startswith('samples.'):
-            ui_facet_name = '%s (samples)' % ui_facet_name
-
-        facets[es_field_name] = {
-            'ui_facet_name': ui_facet_name,
-            'type': field_type
-        }
-        if 'ui_facet_description' in facet_config:
-            facets[es_field_name]['description'] = facet_config[
-                'ui_facet_description']
-
-        facets[es_field_name][
-            'es_facet'] = elasticsearch_util.get_elasticsearch_facet(
-                es, es_field_name, field_type)
+        es_base_field_name = facet_config['elasticsearch_field_name']
+        if elasticsearch_util.is_time_series(es, es_base_field_name):
+            is_time_series = True
+            time_series_vals = elasticsearch_util.get_time_series_vals(
+                es, es_base_field_name, mapping)
+            es_field_names = [
+                es_base_field_name + '.' + tsv for tsv in time_series_vals
+            ]
+        else:
+            is_time_series = False
+            time_series_vals = []
+            es_field_names = [es_base_field_name]
+
+        interval = -1
+        for es_field_name in es_field_names:
+            if es_field_name in facets:
+                raise EnvironmentError('%s appears more than once in ui.json' %
+                                       es_field_name)
+            field_type = elasticsearch_util.get_field_type(es, es_field_name)
+            ui_facet_name = facet_config['ui_facet_name']
+            if es_field_name.startswith('samples.'):
+                ui_facet_name = '%s (samples)' % ui_facet_name
+            facets[es_field_name] = {
+                'ui_facet_name': ui_facet_name,
+                'type': field_type,
+                'is_time_series': is_time_series
+            }
+            if 'ui_facet_description' in facet_config:
+                facets[es_field_name]['description'] = facet_config[
+                    'ui_facet_description']
+
+            if field_type != 'text' and field_type != 'boolean' and interval < 0:
```

Comment:
The one issue with calling get_bucket_interval in get_elasticsearch_facet in the current setup is that for time series facets, get_bucket_interval (which makes a call to elasticsearch) would then be called redundantly for each possible time, which could be inefficient. Maybe it would be cleaner to have a wrapper function get_time_series_elasticsearch_facets to deal with this?

Reply:
Can you try the readable/inefficient approach and see if it's slower? If you don't notice a performance difference, I prefer more readable. (Or if you can think of a readable, efficient approach, that would be fine too.)

Also FYI, the idiomatic Python way of `interval = -1` is `interval = None`/`if not  interval`.